lib/plaid.ex

```diff
@@ -211,23 +65,63 @@ defmodule Plaid do
     end
   end
 
-  defp get_public_key(config) do
-    case Map.get(config, :public_key) || Application.get_env(:plaid, :public_key) do
-      nil ->
-        raise MissingPublicKeyError
+  @doc """
+  Make HTTP request to Plaid.
+  """
+  @callback make_request(atom, String.t(), map, map) ::
+              {:ok, PlaidHTTP.Response.t()} | {:error, PlaidHTTP.Error.t()}
+  def make_request(method, endpoint, parameters, config \\ %{}) do
```

Comment:
`make_request` is a bit misleading since it is making `and` sending the request.

What you could do is to standardized around your data structure as the request and the callback is more like `@callback execute_request(request, config)`

For example, here is a version of it

```ex
defmodule Plaid.Client.Request do
  alias Plaid.Client.Request

  defstruct [:body, :method, :url]

  def new do
    %Request{}
  end

  def put_method(request, m) do
    Map.put(request, :method, m)
  end

  def put_url(request, u) do
    Map.put(request, :url, u)
  end

  def put_body(request, value) do
    Map.put(request, :body, value)
  end

  def merge_body(request, value) do
    request
    |> Map.get(:body)
    |> Map.merge(value)
    |> then(&Map.put(request, :body, &1))
  end

  def to_keyword(request) do
    [body: request.body, url: request.url, method: request.method]
  end
end
```

Here is an example usage

```elixir
def link_token_create(client, body) do
    Request.new()
    |> Request.put_method(:post)
    |> Request.put_url("/link/token/create")
    |> Request.put_body(body)
    |> Client.send_request(client)
  end
```

Reply:
I think the word "make" in `make_request` can be conflated with "build", so I see your point. I like clarity 🙂 

Let me have a think on this and rework it.

Thank you for all the thoughtful feedback!

Reply:
I thought about your feedback here and while I do agree it provides more clarity, it would require a major refactor of the library without much functional benefit aside from a stylistic improvement.

I think about the current structure in terms of responsibilities and testing. Each of the domain modules, e.g. `Plaid.Item`, `Plaid.Accounts` is responsible for the following:

- Knowing the right Plaid endpoint to talk to
- Knowing the right HTTP request to make
- Knowing whether the request requires authentication
- Knowing which data structure to map the response to

The main module, `Plaid`, is responsible for the window dressing common for each request, such as:
- Generating the full url based on the environment and endpoint
- Adding the correct headers
- Handling compile time vs. runtime options
- Merging in static metadata

The HTTP client module, `Plaid.HTTPClient` is responsible for making the actual HTTP request

With the current structure, each module's responsibilities are precise can be easily tested because they're isolated. I can mock the main module to test each domain module's business logic, and mock the HTTP client to test the business logic in the main module, etc.

I'm not saying your suggestion can't accommodate this separation of concerns. In fact, breaking down the responsibilities of `Plaid.make_request/6` into `put_method/2`, `put_url/2`, `put_body/2` allows for precise component testing and it's certainly easier to read and follow, but it would require a lot of re-writing of tests.

Reply:
> The HTTP client module, Plaid.HTTPClient is responsible for making the actual HTTP request

The issue I find with that one is that you are adding another abstraction that Tesla has already given you. Use Tesla Adapter for that.

Leaving you behind with two things to worry about (or maybe one)

1. (Maybe) An abstraction that is just a nicer way to write endpoints so when you read each operation they are as declarative as it gets. That is my `Plaid.Client.Request` and the example code I shared. Otherwise, you can work with a keyword list and remove the module and just work with a bunch of functions that work with keywordlists.
2. Marshal to the right `structured response`. (I haven't done this)

> but it would require a lot of re-writing of tests.

Let me give you a hand, Plaid is not going anywhere for me and your success will be my success as well.

Also, I have been thinking in codegen the client as well, I have experience using OpenAPI spec and Generators around it (https://github.com/straw-hat-team/nodejs-monorepo/tree/master/packages/%40straw-hat/openapi-web-sdk-generator) including how to generate some typing (at least I did it for TypeScript which is more complex than Elixir for sure)

The only thing is that I haven't figure out the most declarative or nicer way to use Jason, so I can generate the parser somehow taking into consideration the struct type. If you find a clean solution it would help a lot.


I can give it a try and see what comes out of it, what I care about the most is that I don't yet publish another client when I can allocate my energy to help you.

Reply:
I'm glad to have your collaboration and I do like your more declarative structure. Let me rework things and push a commit to see if it makes sense to you. 

I admit I had an inkling that I had one layer of abstraction too many here.

I'll also give some thought to a solution for handling the structured response. I have some energy to dedicate to this project right now, so the timing is right.